codegen/smithy-aws-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/aws/protocols/Rpcv2Cbor.kt

```diff
@@ -0,0 +1,150 @@
+        val eventStreamsAcceptHeaderMiddleware = object : ProtocolMiddleware {
+            private val mutateHeadersMiddleware = MutateHeadersMiddleware(extraHeaders = mapOf("Accept" to "application/vnd.amazon.eventstream"))
+
+            override fun isEnabledFor(ctx: ProtocolGenerator.GenerationContext, op: OperationShape): Boolean = op.isOutputEventStream(ctx.model)
+            override val name: String = "Rpcv2CborEventStreamsAcceptHeaderMiddleware"
+            override fun render(ctx: ProtocolGenerator.GenerationContext, op: OperationShape, writer: KotlinWriter) = mutateHeadersMiddleware.render(ctx, op, writer)
+        }
+
+        // Emit a metric to track usage of Rpcv2Cbor
+        val businessMetricsMiddleware = object : ProtocolMiddleware {
+            override val name: String = "Rpcv2CborBusinessMetricsMiddleware"
+            override fun render(ctx: ProtocolGenerator.GenerationContext, op: OperationShape, writer: KotlinWriter) {
+                writer.write("op.context.#T(#T.PROTOCOL_RPC_V2_CBOR)", RuntimeTypes.Core.BusinessMetrics.emitBusinessMetric, RuntimeTypes.Core.BusinessMetrics.SmithyBusinessMetric)
+            }
+        }
+
+        return super.getDefaultHttpMiddleware(ctx) + listOf(
+            smithyProtocolHeaderMiddleware,
+            eventStreamsAcceptHeaderMiddleware,
+            businessMetricsMiddleware,
+        )
+    }
+
+    /**
+     * Exact copy of [AwsHttpBindingProtocolGenerator.renderSerializeHttpBody] but with a custom
+     * [OperationShape.hasHttpBody] function to handle protocol-specific serialization rules.
+     */
+    override fun renderSerializeHttpBody(
+        ctx: ProtocolGenerator.GenerationContext,
+        op: OperationShape,
+        writer: KotlinWriter,
+    ) {
+        val resolver = getProtocolHttpBindingResolver(ctx.model, ctx.service)
+        if (!op.hasHttpBody(ctx)) return
+
+        // payload member(s)
+        val requestBindings = resolver.requestBindings(op)
+        val httpPayload = requestBindings.firstOrNull { it.location == HttpBinding.Location.PAYLOAD }
+        if (httpPayload != null) {
+            renderExplicitHttpPayloadSerializer(ctx, httpPayload, writer)
+        } else {
+            val documentMembers = requestBindings.filterDocumentBoundMembers()
+            // Unbound document members that should be serialized into the document format for the protocol.
+            // delegate to the generate operation body serializer function
+            val sdg = structuredDataSerializer(ctx)
+            val opBodySerializerFn = sdg.operationSerializer(ctx, op, documentMembers)
+            writer.write("val payload = #T(context, input)", opBodySerializerFn)
+            writer.write("builder.body = #T.fromBytes(payload)", RuntimeTypes.Http.HttpBody)
+        }
+        renderContentTypeHeader(ctx, op, writer, resolver)
+    }
```

Comment:
The existing `hasHttpBody` returns `true` if there are any request members bound to the payload. The rpc-v2-cbor spec says:

> Requests for operations with no defined input type (as in, they target the `Unit` shape) MUST NOT contain bodies in their HTTP requests. The Content-Type for the serialization format MUST NOT be set.
 
This difference was causing some protocol tests to fail. For example, when an input targets a non-Unit shape which doesn't have any request members, the original `hasHttpBody` would return `false`, but the CBOR spec expects `true`.

Reply:
Yes that's correct, non-CBOR protocols would have no body in that example, but CBOR *is* expected to send a body (specifically, a `Content-Type` header which is controlled by `hasHttpBody`). 

Here is the specific protocol test that fails without this custom `hasHttpBody` function: https://github.com/smithy-lang/smithy/blob/main/smithy-protocol-tests/model/rpcv2Cbor/empty-input-output.smithy#L106

It fails because there is no `Content-Type` header sent, but it is required